src/main/java/com/gtnewhorizon/gtnhlib/util/data/ItemId.java

```diff
@@ -0,0 +1,132 @@
+package com.gtnewhorizon.gtnhlib.util.data;
+
+import java.util.Objects;
+
+import javax.annotation.Nullable;
+
+import net.minecraft.init.Items;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraftforge.common.util.Constants;
+import net.minecraftforge.oredict.OreDictionary;
+
+import org.jetbrains.annotations.NotNull;
+
+import com.github.bsideup.jabel.Desugar;
+import com.gtnewhorizon.gtnhlib.hash.Fnv1a32;
+
+import it.unimi.dsi.fastutil.Hash;
+
+@SuppressWarnings("unused")
+@Desugar
+public record ItemId(Item item, int meta, NBTTagCompound tag) implements ImmutableItemMeta {
+
+    @Override
+    public @NotNull Item item() {
+        return item;
+    }
+
+    @Override
+    public int getItemMeta() {
+        return meta;
+    }
+
+    @Override
+    public ItemStack toStack(int amount) {
+        ItemStack stack = ImmutableItemMeta.super.toStack(amount);
+
+        if (tag != null) {
+            stack.setTagCompound((NBTTagCompound) tag.copy());
+        }
+
+        return stack;
+    }
+
+    public static ItemId create(NBTTagCompound tag) {
+        return new ItemId(
+                Item.getItemById(tag.getInteger("item")),
+                tag.getInteger("meta"),
+                tag.hasKey("tag", Constants.NBT.TAG_COMPOUND) ? tag.getCompoundTag("tag") : null);
+    }
+
+    public static ItemId create(ItemStack stack) {
+        return create(stack.getItem(), Items.feather.getDamage(stack), stack.getTagCompound());
+    }
+
+    public static ItemId create(Item item, int metaData, @Nullable NBTTagCompound tag) {
+        if (tag != null) {
+            tag = (NBTTagCompound) tag.copy();
+        }
+        return new ItemId(item, metaData, tag);
+    }
+
+    public static ItemId createAsWildcard(ItemStack stack) {
+        return new ItemId(stack.getItem(), OreDictionary.WILDCARD_VALUE, null);
+    }
+
+    public static ItemId createAsWildcardWithNBT(ItemStack stack) {
+        return create(stack.getItem(), OreDictionary.WILDCARD_VALUE, stack.getTagCompound());
+    }
+
+    public static ItemId createWithoutNBT(ItemStack stack) {
+        return new ItemId(stack.getItem(), Items.feather.getDamage(stack), null);
+    }
```

Comment:
Why does createAsWilcard default to "without NBT" but "create" default to "with NBT"? It's a bit confusing. Maybe only have WithNBT and WithoutNBT methods? Then it can't accidentally go wrong.